### PR TITLE
Install Node Problem Detector on GCE Windows nodes

### PR DESCRIPTION
**Pull request by jeremyje**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Adds an option to install Node Problem Detector (NPD) for GCE Windows nodes.
To install NPD for Windows, `WINDOWS_ENABLE_NODE_PROBLEM_DETECTOR=standalone` must set in kubeenv. DaemonSet is not supported due to lack of support of privileged containers.

NPD for Windows is currently pre-alpha but in active development. See https://github.com/kubernetes/node-problem-detector/issues/461.

#### Which issue(s) this PR fixes:

N/A

#### Special notes for your reviewer:
This PR is ready for review but cannot be submitted until the next NPD release that includes Windows tarball artifacts, https://github.com/kubernetes/node-problem-detector/pull/545.

#### Does this PR introduce a user-facing change?

```release-note
Node Problem Detector is now available for GCE Windows nodes.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [Design]: https://docs.google.com/document/d/1eiK6KAp_TFR0PgBMu2WCf49fMZcg-HHnBHMc9fALquU/edit
- [Usage]: https://github.com/kubernetes/node-problem-detector#windows
- [Issue]: https://github.com/kubernetes/node-problem-detector/issues/461
```

/sig node
/sig windows
/priority important-soon

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -1591,9 +1591,18 @@ BOOTSTRAP_KUBECONFIG_FILE: $(yaml-quote "${WINDOWS_BOOTSTRAP_KUBECONFIG_FILE}")
...
+NODE_PROBLEM_DETECTOR_VERSION: $(yaml-quote "${NODE_PROBLEM_DETECTOR_VERSION}")
+NODE_PROBLEM_DETECTOR_TAR_HASH: $(yaml-quote "${NODE_PROBLEM_DETECTOR_TAR_HASH}")
+NODE_PROBLEM_DETECTOR_RELEASE_PATH: $(yaml-quote "${NODE_PROBLEM_DETECTOR_RELEASE_PATH}")
+NODE_PROBLEM_DETECTOR_CUSTOM_FLAGS: $(yaml-quote "${NODE_PROBLEM_DETECTOR_CUSTOM_FLAGS}")
+NODE_PROBLEM_DETECTOR_TOKEN: $(yaml-quote "${NODE_PROBLEM_DETECTOR_TOKEN:-}")
+WINDOWS_NODEPROBLEMDETECTOR_KUBECONFIG_FILE: $(yaml-quote "${WINDOWS_NODEPROBLEMDETECTOR_KUBECONFIG_FILE}")
 EOF
 }
 
+
```
Comment (ibabou) [inline cluster/gce/util.sh:1605]: nit: extra line

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 2]
```diff
@@ -1650,6 +1652,148 @@ function Install-Pigz {
...
+
+      if (${env:CONTAINER_RUNTIME} -eq "containerd") {
+        $system_log_monitors += @("${npd_dir}\config\windows-containerd-monitor-filelog.json")
+        $custom_plugin_monitors += @("${npd_dir}\config\windows-health-checker-containerd.json")
+      } else {
+        $system_log_monitors += @("${npd_dir}\config\windows-containerd-monitor-filelog.json")
+        $custom_plugin_monitors += @("${npd_dir}\config\windows-health-checker-docker.json")
+      }
+
+      # TODO: DO NOT SUBMIT, verify if NPD emits logs to Windows Event Log or should they route to a log file.
```
Comment (ibabou) [inline cluster/gce/windows/k8s-node-setup.psm1:1771]: if this verified, should the "DO NOT SUBMIT" comment be removed ?

**Review by ibabou — APPROVED** [review]

**Comment by pjh** [comment]
cc @mcshooter

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 3]
```diff
@@ -1650,6 +1671,124 @@ function Install-Pigz {
...
+      Log-Output "Creating service: ${NPD_SERVICE}"
+      Log-Output "${npd_bin} ${flags}"
+      sc.exe create $NPD_SERVICE binpath= "${npd_bin} ${flags}" displayName= "Node Problem Detector"
+      sc.exe failure $NPD_SERVICE reset= 30 actions= restart/5000
+      sc.exe start $NPD_SERVICE
+    }
+    else {
+      Log-Output "${NPD_SERVICE} already configured."
+    }
+  }
```
Comment (pjh) [inline cluster/gce/windows/k8s-node-setup.psm1:1789]: Call `Write-VerboseServiceInfoToConsole $NPD_SERVICE` here?

**Comment by pjh** [comment]
/lgtm

Thanks Jeremy!

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 4]
```diff
@@ -1650,6 +1671,124 @@ function Install-Pigz {
   }
 }
 
+# Node Problem Detector Resources
+$NPD_SERVICE = "node-problem-detector"
+$DEFAULT_NPD_VERSION = '0.8.7-17-g06732e5'
+# DO NOT SUBMIT - Remove the jeremyje-public from RELEASE_PATH and use an NPD release that includes Windows support.
+$DEFAULT_NPD_RELEASE_PATH = 'https://storage.googleapis.com/jeremyje-public/kubernetes-release'
```
Comment (mcshooter) [inline cluster/gce/windows/k8s-node-setup.psm1:1678]: do you want to update these paths to nodeproblemdetector-artifacts?

Comment (jeremyje) [reply cluster/gce/windows/k8s-node-setup.psm1:1678]: It's best to wait for NPD to update their build first.

File: cluster/gce/util.sh [hunk 5]
```diff
@@ -1591,6 +1591,13 @@ BOOTSTRAP_KUBECONFIG_FILE: $(yaml-quote "${WINDOWS_BOOTSTRAP_KUBECONFIG_FILE}")
 KUBEPROXY_KUBECONFIG_FILE: $(yaml-quote "${WINDOWS_KUBEPROXY_KUBECONFIG_FILE}")
 WINDOWS_INFRA_CONTAINER: $(yaml-quote "${WINDOWS_INFRA_CONTAINER}")
 WINDOWS_ENABLE_PIGZ: $(yaml-quote "${WINDOWS_ENABLE_PIGZ}")
+ENABLE_NODE_PROBLEM_DETECTOR: $(yaml-quote "${WINDOWS_ENABLE_NODE_PROBLEM_DETECTOR}")
+NODE_PROBLEM_DETECTOR_VERSION: $(yaml-quote "${NODE_PROBLEM_DETECTOR_VERSION}")
+NODE_PROBLEM_DETECTOR_TAR_HASH: $(yaml-quote "${NODE_PROBLEM_DETECTOR_TAR_HASH}")
+NODE_PROBLEM_DETECTOR_RELEASE_PATH: $(yaml-quote "${NODE_PROBLEM_DETECTOR_RELEASE_PATH}")
```
[the remark this answers is not in the document]
Comment (jeremyje) [reply cluster/gce/util.sh:1597]: Yes these are NPD parameters that are set in other parts of the kube-up.

**Review by mcshooter — APPROVED** [review]
Lgtm

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 4 above)
Comment (jeremyje) [reply cluster/gce/windows/k8s-node-setup.psm1]: NPD 0.8.8 has been released with Windows support! Updated the references.

**Pull request merged**